tools/topology/platform/intel/intel-generic-dmic-kwd.m4

```diff
@@ -0,0 +1,123 @@
+
+# variable that need to be defined in upper m4
+# define(CHANNELS, `4')
+
+# define(KFBM_TYPE, `vol-kfbm')
+
+# define(DMIC_PCM_48k_ID, `99')
+# define(DMIC_PIPELINE_48k_ID, `4')
+# define(DMIC_DAI_LINK_48k_ID, `1')
+
+# define(DMIC_PCM_16k_ID, `7')
+# define(DMIC_PIPELINE_16k_ID, `9')
+# define(DMIC_PIPELINE_KWD_ID, `10')
+# define(DMIC_DAI_LINK_16k_ID, `2')
+
+# define(KWD_PIPE_SCH_DEADLINE_US, 20000)
+
+#
+# Define the pipelines
+#
+
+dnl PIPELINE_PCM_ADD(pipeline,
+dnl     pipe id, pcm, max channels, format,
+dnl     period, priority, core,
+dnl     pcm_min_rate, pcm_max_rate, pipeline_rate,
+dnl     time_domain, sched_comp)
+
+# Passthrough capture pipeline using max channels defined by CHANNELS.
+
+# Set 1000us deadline on core 0 with priority 0
+PIPELINE_PCM_ADD(sof/pipe-passthrough-capture.m4,
+	DMIC_PIPELINE_48k_ID, DMIC_PCM_48k_ID, CHANNELS, s32le,
+	1000, 0, 0, 48000, 48000, 48000)
+
+#
+# KWD configuration
+#
+
+# Passthrough capture pipeline 7 on PCM 7 using max 2 channels.
+# Schedule 20000us deadline on core 0 with priority 0
+PIPELINE_PCM_DAI_ADD(sof/pipe-KFBM_TYPE-capture.m4,
+	DMIC_PIPELINE_16k_ID, DMIC_PCM_16k_ID, 2, s24le,
+	KWD_PIPE_SCH_DEADLINE_US, 0, 0, DMIC, 1, s32le, 3,
+	16000, 16000, 16000)
+
+
+
+#
+# DAIs configuration
+#
+
+dnl DAI_ADD(pipeline,
+dnl     pipe id, dai type, dai_index, dai_be,
+dnl     buffer, periods, format,
+dnl     deadline, priority, core, time_domain)
+
+# capture DAI is DMIC 0 using 2 periods
+# Buffers use s32le format, 1000us deadline on core 0 with priority 0
+DAI_ADD(sof/pipe-dai-capture.m4,
+	DMIC_PIPELINE_48k_ID, DMIC, 0, dmic01,
+	concat(`PIPELINE_SINK_', DMIC_PIPELINE_48k_ID), 2, s32le,
+	1000, 0, 0, SCHEDULE_TIME_DOMAIN_TIMER)
+
+# capture DAI is DMIC 1 using 3 periods
+# Buffers use s32le format, with 320 frame per 20000us on core 0 with priority 0
+DAI_ADD(sof/pipe-dai-capture.m4,
+	DMIC_PIPELINE_16k_ID, DMIC, 1, dmic16k,
+	`PIPELINE_SINK_'DMIC_PIPELINE_16k_ID, 3, s32le,
+	KWD_PIPE_SCH_DEADLINE_US, 0, 0, SCHEDULE_TIME_DOMAIN_TIMER)
+
+dnl PCM_DUPLEX_ADD(name, pcm_id, playback, capture)
+dnl PCM_CAPTURE_ADD(name, pipeline, capture)
+PCM_CAPTURE_ADD(DMIC48kHz, DMIC_PCM_48k_ID, concat(`PIPELINE_PCM_', DMIC_PIPELINE_48k_ID))
+
+# keyword detector pipe
+dnl PIPELINE_ADD(pipeline,
+dnl     pipe id, max channels, format,
+dnl     period, priority, core,
+dnl     sched_comp, time_domain,
+dnl     pcm_min_rate, pcm_max_rate, pipeline_rate)
+PIPELINE_ADD(sof/pipe-detect.m4,
+	DMIC_PIPELINE_KWD_ID, 2, s24le,
+	KWD_PIPE_SCH_DEADLINE_US, 1, 0,
+	`PIPELINE_SCHED_COMP_'DMIC_PIPELINE_16k_ID,
+	SCHEDULE_TIME_DOMAIN_TIMER,
+	16000, 16000, 16000)
+
+# Connect pipelines together
+SectionGraph."pipe-sof-generic-keyword-detect" {
+        index "0"
+
+        lines [
+		# keyword detect
+          dapm(`PIPELINE_SINK_'DMIC_PIPELINE_KWD_ID, `PIPELINE_SOURCE_'DMIC_PIPELINE_16k_ID)
+					dapm(`PIPELINE_PCM_'DMIC_PIPELINE_16k_ID, `PIPELINE_DETECT_'DMIC_PIPELINE_KWD_ID)
+        ]
```

Comment:
something wonky with alignment here |?

Reply:
This is some copy paste from some other files, what should we use in M4 file? Tab or space?
Will try to check the alignment.